### ISSUE TITLE
Input plugin does not expose port on arbitrary node

### ISSUE THREAD

**Issue by tetramin** [fluent/fluent-plugin-prometheus#92]
I'm starting fluentd with prometheus plugin as a daemonset in GKE. Often, when the container is started, port 24231 does not open, although on other nodes the ports are normally exposed. When I enter the `ss -lnt` command, I do not see open ports in container.

There are no suspicious messages in the logs even in trace mode.

How I can debug this?

**Comment by SergeyD** [comment]
We have the same issue on k8s clusters running on bare-metal instances: 7 of 54 containers do not listening on port without any log messages (`debug` level used).

**Comment by SergeyD** [comment]
I did some experiments by editing fluentd configuration file directly in running containers and applying changes by sending `HUP` signal to fluentd process to re-load config.

Two observations:
1. Disabling `in_tail` input (targeting k8s container) logs resolves issue with binding after config reload.
2. With enabled `in_tail` and elasticsearch output replaced with 
```
<match kube.**>
        @type null 
</match>
``` 
plugin was able to bind after a short time. 
After this I reverted output back to elasticsearch plugin, reloaded config - and plugin still able to bind.

Looks like big input queue/pressure prevents plugin from bind on startup time.

**Comment by SergeyD** [comment]
I've find issue and solution for my case. We have two inputs configured for our `fluentd` instances. 
One of them is `in_tail` for k8s logs:

```
<source>
      @type tail
      path /var/log/containers/*.log
      pos_file /var/log/es-containers.log.pos
      tag kubernetes.*
      format none
      refresh_interval 60
    </source>
```

From startup logs in `debug` mode it seen that `in_tail` found around 80 tailing paths separated by comma:

```
2019-05-25 16:24:34 +0000 [debug]: #0 tailing paths: target = /var/log/containers/XXX.log,/var/log/containers/YYY.log,...,....```

Next message is:

```2019-05-25 16:24:34 +0000 [info]: #0 following tail of /var/log/containers/XXX.log```

And no more `following tail of ...` messages, so obviously it stuck on file `XXX.log` which is around 7Gb in my case.

I've added `skip_refresh_on_startup true` setting for `in_tail` plugin, and now monitoring http source able to bind on startup.

**Comment by repeatedly** [comment]
Yes. in_tail launches and starts file watchers during startup phase by default.
This is no problem for almost cases but if you have large files, it takes long time to process it.
`skip_refresh_on_startup` avoids this problem by disabling watcher launch at startup.

**Comment by Watson1978** [comment]
Thanks for the detailed investigation.
Your analysis is correct, and this is expected Fluentd startup behavior rather than a bug in fluent-plugin-prometheus itself.

Root cause: within a single worker, Fluentd starts input plugins sequentially, and `in_tail` performs its initial watch-list refresh synchronously during the startup phase by default. 
When one of the tailed files is very large (your 7 GB `XXX.log`), that initial read blocks the rest of the startup sequence, so `in_prometheus` never reaches the point where it binds its HTTP server on 24231. That is exactly why `ss -lnt` shows no listening socket and there is nothing in the logs even at trace level — the process is simply still busy reading the file, not erroring.

This is documented behavior of `in_tail`: by default it reads the target files during startup, and if a file is large enough that this read takes a while, the startup of other plugins is blocked until it finishes. The docs explicitly recommend `skip_refresh_on_startup` to avoid it (see https://docs.fluentd.org/input/tail).

So the fix you already found is the correct one:

```
<source>
  @type tail
  path /var/log/containers/*.log
  pos_file /var/log/es-containers.log.pos
  tag kubernetes.*
  skip_refresh_on_startup true
  refresh_interval 60
</source>
```

A couple of options if you want more isolation of the metrics endpoint:
- Run `in_prometheus` in its own worker (via `<system> workers N </system>` + `<worker>` directives) so a busy `in_tail` in another worker cannot delay the metrics port from binding — separate workers are separate processes with independent event loops.
- Keep `in_tail` startup cheap by persisting `pos_file` across restarts and only using `read_from_head` where you actually need it, so it does not re-scan whole files on every start.

Since this is expected Fluentd startup behavior with a working configuration fix rather than a plugin-side bug, I'll close this issue.
Feel free to reopen if the metrics port still fails to bind with `skip_refresh_on_startup true` set.
